crates/sui-core/src/consensus_adapter.rs

```diff
@@ -206,62 +210,76 @@ impl ConsensusAdapter {
         self.num_inflight_transactions.load(Ordering::Relaxed)
     }
 
-    fn should_submit(
+    fn await_submit_delay(
         committee: &Committee,
         ourselves: &AuthorityName,
         transaction: &ConsensusTransaction,
-    ) -> bool {
+    ) -> impl Future<Output = ()> {
+        tokio::time::sleep(Self::submit_delay(committee, ourselves, transaction))
+    }
+
+    fn submit_delay(
+        committee: &Committee,
+        ourselves: &AuthorityName,
+        transaction: &ConsensusTransaction,
+    ) -> Duration {
         if let ConsensusTransactionKind::UserTransaction(certificate) = &transaction.kind {
-            Self::should_submit_certificate(committee, ourselves, certificate.digest())
+            Self::submit_delay_certificate(committee, ourselves, certificate.digest())
         } else {
-            true
+            Duration::ZERO
         }
     }
 
-    /// Check if this authority should submit the certificate to consensus.
-    fn should_submit_certificate(
+    /// Check when this authority should submit the certificate to consensus.
+    fn submit_delay_certificate(
+        committee: &Committee,
+        ourselves: &AuthorityName,
+        tx_digest: &TransactionDigest,
+    ) -> Duration {
+        let (nominator, denominator) =
+            Self::position_submit_certificate(committee, ourselves, tx_digest);
+        const MAX_DELAY_MS: u64 = 60_000;
+        // There is no overflow as long as total_stake < u64::MAX / MAX_DELAY_MS, which seem to be reasonable expectation
+        let duration_ms = MAX_DELAY_MS * nominator / denominator;
+        Duration::from_millis(duration_ms)
+    }
+
+    /// Returns a position of the current validator in ordered list of validator to submit transaction
+    /// You can think of this value as a float in range [0, 1)
+    /// Validators closer to 0 are submitting transaction sooner, and validators closer to 1 wait longer
+    ///
+    /// In practice, instead of returning float we return integer fraction here in form of (nominator, denominator)
+    /// If returned tuple (A, B) the value A is in range [0, B)
+    fn position_submit_certificate(
         committee: &Committee,
         ourselves: &AuthorityName,
         tx_digest: &TransactionDigest,
-    ) -> bool {
+    ) -> (StakeUnit, StakeUnit) {
         // the 32 is as requirement of the deault StdRng::from_seed choice
-        let digest_bytes = if let Some(b) = tx_digest.to_bytes().get(0..32) {
-            if let Ok(bytes) = b.try_into() {
-                bytes
-            } else {
-                return false;
-            }
-        } else {
-            return false;
-        };
+        let digest_bytes = tx_digest.into_bytes();
 
         // permute the validators deterministically, based on the digest
         let mut rng = StdRng::from_seed(digest_bytes);
         let mut validators = committee.voting_rights.clone();
         validators.shuffle(&mut rng);
```

Comment:
well, this is not part of your code, but should we use `committee.shuffle_by_stake()` rather than using an unweighted distribution?

Reply:
We do take weight into account below in this function, so I am not sure why we need to take it into account during shuffling

Reply:
I concur that we should take stake into account in the shuffling, and augment the shuffle to take a seed to allow for the use case here.

Not using the weight leads to the following attack (in theory, might mitigate by having min stake thresholds):
* Adversary introduces a very large number of validators (A = 100s) with very little stake each (say 0.1%/A) and overall (0.1%) in a committee with honest validators H << A.
* Now a random shuffle will likely have an initial run of A / H adversary nodes, which do not include the transaction in the consensus.
* The resulting latency is multiplied by A / H despite the adversary having minute stake.

When shuffling by stake we expect that the run until a good node is selected is 3/2, assuming the adversary has at most 1/3 stake.